codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/endpoints/EndpointParametersGenerator.java

```diff
@@ -0,0 +1,299 @@
+                $parametersDefaultValueMethod:W
+                """,
+                commonCodegenArgs,
+                MapUtils.of(
+                        "parametersTypeDocs", generateParametersTypeDocs()),
+                overriddenArgs);
+    }
+
+    private GoWriter.Writable generateParametersTypeDocs() {
+        return goDocTemplate("""
+                $parametersType:T provides the option parameters for influence how endpoints are resolved.
+                """, commonCodegenArgs);
+    }
+
+    private GoWriter.Writable generateParametersMembers(Parameters parameters) {
+        return (GoWriter w) -> {
+            w.indent();
+            sortParameters(parameters).forEach((parameter) -> {
+                var writeChain = new GoWriter.ChainWritable()
+                        .add(parameter.getDocumentation(), GoWriter::goTemplate)
+                        .add(parameter.isRequired(), goTemplate("Parameter is required."))
+                        .add(parameter.getDefaultValue(), (defaultValue) -> {
+                            return goTemplate("Defaults to " + defaultValue + " if no value is provided.");
+                        })
+                        .add(parameter.getBuiltIn(), GoWriter::goTemplate)
+                        .add(parameter.getDeprecated(), (deprecated) -> {
+                            return goTemplate("Deprecated: " + deprecated.getMessage());
+                        });
+                w.writeRenderedDocs(writeChain.compose());
+                w.write("$L $P", getExportedParameterName(parameter), parameterAsSymbol(parameter));
+                w.write("");
+            });
+
+            w.write("$W", generateLoggerMember());
+        };
+    }
+
+    private GoWriter.Writable generateLoggerMember() {
+        Map<String, Object> loggerArgs = MapUtils.of(
+                "loggerMemberName", LOGGER_MEMBER_NAME,
+                "loggerType", SymbolUtils.createValueSymbolBuilder("Logger",
+                        SmithyGoDependency.SMITHY_LOGGING).build(),
+                "enableLoggingMemberName", ENABLE_LOGGING_MEMBER_NAME);
+        return goTemplate("""
+                $loggerMemberDocs:W
+                $loggerMemberName:L $loggerType:T
+
+                $enableLoggingMemberDocs:W
+                $enableLoggingMemberName:L bool
+                """,
+                loggerArgs,
+                MapUtils.of(
+                        "loggerMemberDocs", goDocTemplate(
+                                "$loggerMemberName:L provides logging of errors occurring when resolving an endpoint.",
+                                loggerArgs),
+                        "enableLoggingMemberDocs", goDocTemplate(
+                                "$loggerMemberName:L provides logging of errors occurring when resolving an endpoint.",
+                                loggerArgs)));
+
+    }
+
+    private GoWriter.Writable generateDefaultsMethod(Parameters parameters) {
```

Comment:
So the doc says
```go
// WithDefaults returns a shallow copy of EndpointParameterswith default values
// applied to members where applicable.
```
But it's not actually copying anything, it's just setting defaults in-place on the receiver and then returning that receiver.

That aside, I wonder if it would be easier to consume down the road in codegen to generate something like `NewWithDefaults()` (that behaves accordingly).

Reply:
hmm so i think its helpful to look at the generated code, from the signature you can see that the receiver is a copy and the return is a copy.
```go
func (p EndpointParameters) WithDefaults() EndpointParameters {
	if p.UseDualStack == nil {
	    p.UseDualStack = ptr.Bool(false)
	}
	
	if p.UseFIPS == nil {
	    p.UseFIPS = ptr.Bool(false)
	}
	return p
}
```

and wrt to `NewWithDefaults`, easier in what sense?

Reply:
oh, so my understanding is that when Go copies value receivers, they are shallow copies: any members which are themselves pointer values only copy the pointer? so for example `p.Region`, if `Region` is a `*string`, the string value wont be copied only the pointer to it.